features/include/pcl/features/impl/moment_of_inertia_estimation.hpp

```diff
@@ -221,52 +213,25 @@ pcl::MomentOfInertiaEstimation<PointT>::getOBB (PointT& min_point, PointT& max_p
 template <typename PointT> void
 pcl::MomentOfInertiaEstimation<PointT>::computeOBB ()
 {
-  obb_min_point_.x = std::numeric_limits <float>::max ();
-  obb_min_point_.y = std::numeric_limits <float>::max ();
-  obb_min_point_.z = std::numeric_limits <float>::max ();
-
-  obb_max_point_.x = std::numeric_limits <float>::min ();
-  obb_max_point_.y = std::numeric_limits <float>::min ();
-  obb_max_point_.z = std::numeric_limits <float>::min ();
-
-  unsigned int number_of_points = static_cast <unsigned int> (indices_->size ());
-  for (unsigned int i_point = 0; i_point < number_of_points; i_point++)
-  {
-    float x = (input_->points[(*indices_)[i_point]].x - mean_value_ (0)) * major_axis_ (0) +
-              (input_->points[(*indices_)[i_point]].y - mean_value_ (1)) * major_axis_ (1) +
-              (input_->points[(*indices_)[i_point]].z - mean_value_ (2)) * major_axis_ (2);
-    float y = (input_->points[(*indices_)[i_point]].x - mean_value_ (0)) * middle_axis_ (0) +
-              (input_->points[(*indices_)[i_point]].y - mean_value_ (1)) * middle_axis_ (1) +
-              (input_->points[(*indices_)[i_point]].z - mean_value_ (2)) * middle_axis_ (2);
-    float z = (input_->points[(*indices_)[i_point]].x - mean_value_ (0)) * minor_axis_ (0) +
-              (input_->points[(*indices_)[i_point]].y - mean_value_ (1)) * minor_axis_ (1) +
-              (input_->points[(*indices_)[i_point]].z - mean_value_ (2)) * minor_axis_ (2);
-
-    if (x <= obb_min_point_.x) obb_min_point_.x = x;
-    if (y <= obb_min_point_.y) obb_min_point_.y = y;
-    if (z <= obb_min_point_.z) obb_min_point_.z = z;
-
-    if (x >= obb_max_point_.x) obb_max_point_.x = x;
-    if (y >= obb_max_point_.y) obb_max_point_.y = y;
-    if (z >= obb_max_point_.z) obb_max_point_.z = z;
-  }
+  obb_min_point_.getArray3fMap () = std::numeric_limits <float>::max ();
+  obb_max_point_.getArray3fMap () = -std::numeric_limits <float>::max ();
 
   obb_rotational_matrix_ << major_axis_ (0), middle_axis_ (0), minor_axis_ (0),
                             major_axis_ (1), middle_axis_ (1), minor_axis_ (1),
                             major_axis_ (2), middle_axis_ (2), minor_axis_ (2);
 
-  Eigen::Vector3f shift (
-    (obb_max_point_.x + obb_min_point_.x) / 2.0f,
-    (obb_max_point_.y + obb_min_point_.y) / 2.0f,
-    (obb_max_point_.z + obb_min_point_.z) / 2.0f);
+  for (const auto& idx: *indices_)
+  {
+    const Eigen::Vector3f diff = input_->points[idx].getVector3fMap () - mean_value_;
```

Comment:
Wouldn't it be better to use `auto` in such a case? I'm not sure whether it applies to that case, but I seem to remember that Eigen uses expression templates and that using `auto` here would allow the compiler to generate better code in the following line thanks to it. I wouldn't push for it without a codegen example though.

Reply:
Currently, I'm debugging a test failure (on and off, can't give sufficient time to it) but some concerns with using `auto` with Eigen are:
* temporaries instead of expression templates (only in some conditions)
* duplicated evaluation due to expressions templates (just a need to be careful before using auto in more than 1 expression)

I'm definitely using `auto` in a lot of other places with Eigen and will try to incorporate your suggestion once the test passes.

2 tests are failing on my setup (haven't checked CI yet though):
* 36 - feature_spin_estimation (Failed)
* 45 - feature_moment_of_inertia_estimation (Failed)

Brisk 2D and Shot are no longer failing.

Reply:
> I wouldn't push for it without a codegen example though.

I created [a small prototype](https://godbolt.org/z/2u1CBY) which shows almost the same assembly. The difference stems from the fact that the auto version doesn't store the difference in a separate variable. Don't think the benchmarks will be that different (I can't benchmark so just speculating)

Reply:
I was mostly curious about whether it was possible to use the quirks of Eigen to gain some speed here and there, but the concerns you have about temporaries and duplicate evaluation are valid. I guess it's only worth it in tight loops where it is shown to make a real difference :)

Reply:
I'm using auto where possible (within the constraints I mentioned). But .. I can't find the bugs. It's a bit frustrating TBH.

In some cases, it doesn't seem like Eigen has special conditions for a complicated expression template of the form `(A-B)*(A-B)^T` (it'll perform subtraction twice in the expression, despite it needing to be done once). I've checked it in the codegen.

Reply:
I've burnt myself quite a few times by using `auto` and Eigen. Nowadays I have a strict rule for myself: never ever use auto` when manipulating Eigen types.

In fact, on their [Common pitfals](https://eigen.tuxfamily.org/dox/TopicPitfalls.html) page the write:

> In short: do not use the auto keywords with Eigen's expressions, unless you are 100% sure about what you are doing. 

Golden words.